### ISSUE TITLE
ol.interaction.Select() ?

### ISSUE THREAD

**Issue by michaelletzgus** [openlayers/openlayers#5831]
Hi All!

I have a few questions|problems|enhancements  with/for the ol.interaction.Select.

Problem: Multiple select interactions do not work - but I have a lot of different modules exporting layers to my main map. A few of these layers need to support selection, but only one select is possible, so this can't be implemented in the different modules.

I tried to create kind generic select mechanisms for the main map, but every solution looks a little bit complicated

1.
Styles. I'm using one stylefunc per layer, so the layers stylefunc must also be the interaction style. The different styles for (non-)selected are all handled in my layers stylefuncs.

Problem: How to supply the layers stylefunc to the interaction?

``` javascript
this.selectInteraction = new ol.interaction.Select({
            layers: [this.ANIMOVERLAY.layer_particle],
            pixelTolerance:30,
            style:function(feature, resolution) {
                return this.selectInteraction.getLayer(feature).getStyleFunction()(feature, resolution)

            }.bind(this),
        })
```

Since the style can't be changed later the function must be supplied in the ctor, so this strange bind() cascade is required.

Idea: Use a "styler" callback function like this:

``` javascript
styler(feature, resolution){
return style or [style] or stylefunc
}
```

Every time the interactions needs to know the style `styler(f,r).bind(this)` is called. This would allow more flexible style functions, knowing everything the interactions knows.

2.
Notification
How does a feature know whether it's selected or not? Currently I'm setting my own selected flag:

``` javascript
this.selectInteraction.on('select',function(evt){
    $.each(evt.selected, function(idx, f) {
        f.set('_selected', true, true)
    })
    $.each(evt.deselected, function(idx, f) {
        f.set('_selected', false, true)
    })
})
```

Is there any way to do this smoother? Maybe firing a 'un-selected' event for the feature?

3.
Interference with map.on('click')

I'm using `map.on('click')` to place markers on my map. If I select a feature the marker is placed on this feature. I don't know ho to prevent this without establishing my own select mechanism inside the click-handler.

Idea:  The `map.on('click')`-handler could append an callback function at the end of the event chain. This callback will only be called if nobody else calls .stopPropagation().

Possible problem: There is a small delay (waiting for dblclick?) when selecting with the select interaction...

Any ideas?

Best regards
 Michael

**Comment by simonseyock** [comment]
To this subject i would like to mention this post https://groups.google.com/forum/#!topic/ol3-dev/Yh7eMfcUFBI and this PR: #5279

**Comment by simonseyock** [comment]
About 2: You can check if the feature is in the collection returned by the getFeatures method of the select interaction.

**Comment by michaelletzgus** [comment]
That's true, but I want the feature only to know whether it's selected or not - not more.
Of course I can supply the selected collection to all style functions and/or features - but this is much more information than necessary. 

I think it would be better to not let the feature now by which mechanisms it had been selected and how this mechanism works. The feature only should now the answer to "Am I selected?" Or, even better, each feature receives an event notification.

**Comment by ahocevar** [comment]
@michaelletzgus It sounds like what you are looking for is not `ol.interaction.Select`, but `ol.Map#forEachFeatureAtPixel`. This allows you to handle things the way you desire, including setting an attribute on a selected feature. To be honest, I have stopped using `ol.interaction.Select` because I find the `forEachFeatureAtPixel` approach simpler, more flexible, less overhead and 100% predictable.

**Comment by michaelletzgus** [comment]
@ahocevar  Thats excatly what I'm doing now, the only problem was that my `on('click')` handler is still using `forEachFeatureAtPixel()` - but for a lot of other things.
I think `ol.interaction.Select` is quite useless, the same feature can be implemented with a few lines of code, performing better and having more features than.
Maybe `ol.interaction.Select` would be more useful or easier to use and more intuitive if it fires select and unselect events on the features?

**Comment by ahocevar** [comment]
Maybe. I'll leave that to others to decide and implement. As I said [elsewhere](5255), I'm done with the Select interaction 😄.

**Comment by michaelletzgus** [comment]
Live is great without `ol.Interaction.Select` :)
Is it possible to achieve some kind of "per geometry" of "per style" filtering with `forEachFeatureAtPixel`?
I have a lot of moving Point features, styled by the Layers stylefunc. The style adds a Text style and two additional geometries (a heading indicator line and a trail) to the point.
Is it possible to only retrieve clicks on the point - but not on the text nor the additional geometries?
I tried to find a solution in the ol3 source - but I got lost somewhere in `ol.structs.RBush.prototype.getInExtent = function(extent)` in rbush.js.

My current solution is an additional distance of click check in my features event handler, this works from the point of view of the feature. But the event dispatching `forEachFeatureAtPixel`-loop doesn't know anything about the result of this check.
Maybe I should not use `dispatchEvent()` but mym own event dispatch function with an appropriate  return value...?
